rskj-core/src/main/java/org/ethereum/util/MapSnapshot.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2021 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.util;
+
+import org.ethereum.db.ByteArrayWrapper;
+
+import javax.annotation.Nonnull;
+import java.io.*;
+import java.util.Map;
+import java.util.Objects;
+
+public abstract class MapSnapshot<T extends Closeable> implements Closeable {
+
+    protected final T stream;
+
+    protected MapSnapshot(@Nonnull T stream) {
+        this.stream = Objects.requireNonNull(stream);
+    }
+
+    @Override
+    public void close() throws IOException {
+        stream.close();
+    }
+
+    public interface Factory {
+
+        default MapSnapshot.In makeInputSnapshot(@Nonnull InputStream inputStream) {
+            return new MapSnapshot.In(new DataInputStream(inputStream));
+        }
+
+        default MapSnapshot.Out makeOutputSnapshot(@Nonnull OutputStream outputStream) {
+            return new MapSnapshot.Out(new DataOutputStream(outputStream));
+        }
+
+    }
+
+    public static class In extends MapSnapshot<DataInputStream> {
+
+        protected In(@Nonnull InputStream input) {
+            super(new DataInputStream(input));
+        }
+
+        public void read(@Nonnull Map<ByteArrayWrapper, byte[]> map) throws IOException {
```

Comment:
I think we can add some javadoc explaining the order in what the values are read and, for example, the constraints or possible values (keySize should be >= 1, valueSize should be >= -1, etc). I think it could help to understand the code better besides it is really tidy and easy to read, but is not that clear what's happening without having to read the whole function beforehand.